### PR TITLE
fix simd op check vdmpy and vrmpy failures

### PR DESCRIPTION
**Pull request by pranavb-ca**
On simd_op_check I was seeing two failures -  Unable to match vdmpy and vrmpy. This is mainly because the new simplifier and GroupLoopInvariants reassociate some add expressions causing them to fall through the cracks in our pattern matching logic.

@dsharletg, @aankit-ca, @dpalermo, @suyogsarda  - PTAL.

### CODE REVIEW

File: src/HexagonOptimize.cpp [hunk 1]
```diff
@@ -490,7 +490,62 @@ class OptimizePatterns : public IRMutator2 {
...
+        const Shuffle *first_shuffle = mpys[0].first.as<Shuffle>();
+        if (first_shuffle) {
+            for (MulExpr &m : mpys) {
+                const Shuffle *shuffle = m.first.as<Shuffle>();
+                if (!shuffle || !shuffle->is_slice()) {
+                    return;
+                }
+            }
+            std::stable_sort(mpys.begin(), mpys.end(),
+                             [&](const MulExpr &m1, const MulExpr &m2) {
```
[the remark this answers is not in the document]
Comment (pranavb-ca) [reply src/HexagonOptimize.cpp]: aah, nice catch. Fixed.

File: src/HexagonOptimize.cpp [hunk 2]
```diff
@@ -490,7 +490,62 @@ class OptimizePatterns : public IRMutator2 {
...
-
+    // We'll try to sort the mpys based my mpys.first.
+    // But, for this all the mpy.first exprs should either be
+    // all loads or all slice_vectors.
+    static void sort_mpy_exprs(vector<MulExpr> &mpys) {
+        struct LoadCompare {
+            Expr ramp_base;
+            LoadCompare(Expr e) : ramp_base(e) {}
+            bool operator()(const MulExpr &m1, const MulExpr &m2) {
+                if (!m1.first.defined() || !m2.first.defined())
```
[the remark this answers is not in the document]
Comment (pranavb-ca) [reply src/HexagonOptimize.cpp]: Fixed.

**Comment by steven-johnson** [comment]
LGTM, but should get @dsharletg approval to land.

File: src/HexagonOptimize.cpp [hunk 3]
```diff
@@ -490,7 +490,62 @@ class OptimizePatterns : public IRMutator2 {
...
+        if (first_shuffle) {
+            for (MulExpr &m : mpys) {
+                const Shuffle *shuffle = m.first.as<Shuffle>();
+                if (!shuffle || !shuffle->is_slice()) {
+                    return;
+                }
+            }
+            std::stable_sort(mpys.begin(), mpys.end(),
+                             [&](const MulExpr &m1, const MulExpr &m2) {
+                                 return m1.first.as<Shuffle>()->slice_begin() < m2.first.as<Shuffle>()->slice_begin();
```
[the remark this answers is not in the document]
Comment (pranavb-ca) [reply src/HexagonOptimize.cpp]: Yes, left it to the interleaver

File: src/HexagonOptimize.cpp [hunk 4]
```diff
@@ -490,7 +490,62 @@ class OptimizePatterns : public IRMutator2 {
...
+            bool operator()(const MulExpr &m1, const MulExpr &m2) {
+                if (!m1.first.defined() || !m2.first.defined())
+                    return false;
+                const Load *m1_load = m1.first.as<Load>();
+                const Load *m2_load = m2.first.as<Load>();
+                internal_assert(m1_load && m2_load);
+                const Ramp *m1_ramp = m1_load->index.as<Ramp>();
+                const Ramp *m2_ramp = m2_load->index.as<Ramp>();
+                internal_assert(m1_ramp && m2_ramp);
+                Expr diff_m1 = simplify(m1_ramp->base - ramp_base);
```
[the remark this answers is not in the document]
Comment (pranavb-ca) [reply src/HexagonOptimize.cpp]: makes sense. will do

**Comment by pranavb-ca** [comment]
@dsharletg - The two buildbot failures are unrelated.  linux-64-gcc53-testbranch-trunk which tests HVX is green. Ok to merge?

**Pull request merged**